### ISSUE TITLE
Load BS 3 javascript file

### ISSUE THREAD

**Issue by laoneo** [Digital-Peak-Incubator/tpl_tauristar#87]
When bs3 is defined in the template, the bs3 javascript files should be loaded in JHtmlBootstrap.

**Comment by dneukirchen** [comment]
Any ideas on how to accomplish this?

The easiest solution would be to modify the signature of JHtmlBootstrap::framework($debug = null) to JHtmlBootstrap::framework($debug = null, $version = 2) and pass the correct version from the template. But this means (if we dont wanna misuse the static::$loaded cache), that we have to modifiy all JHtml::_('bootstrap.framework') calls.

Another idea is to modify JHtml::_('script', ...) like we did it with template files (bootstrap.bs3.min.js), but this doesnt make much sense in my opinion.

A third idea is to wrap the current JHtmlBootstrap class into a VersionAwareJHtmlBootstrap class wrapper, which initializes with the correct framework version. 

Last but not least we could create a JHtmlBootstrap3 class and make the switch on the template level, but this also means, that modifications of JHtml::_('framework.bootstrap') calls are necessary.

Any thoughts/ideas?

**Comment by yvesh** [comment]
All ideas have up and downs, that was the reason i suggested setting bstrap version in the global configuration, as you can then use JFactory::getApplication()->get('bootstrapVersion').

What we also could do is extending the joomla config (JRegistry (long ago getConfig()) with an template configuration on the initialization (would offer more flexibility for templates too)..

**Comment by laoneo** [comment]
JApplicationHelper has a new function to get the frontend framework (ff) which should be loaded. I would use this one within a new class JHtmlFrontendFramework do determine the ff to load. JHtmlBootstrap should be deprecated then. Passing the ff as version parameter I wouldn't do, as extension devs should not care about which ff to load.

The parameter which sets the ff should be in the template configuration as it can be, to have a bs2 and bs3 template on the same site.

**Comment by laoneo** [comment]
I'm on my mobile, but I think there is already something like jhtmlframework.

**Comment by laoneo** [comment]
I would set the frongend framework (ff) parameter in the template, as it is possible to have a bs2 and bs 3 template on the same site, assigned to differdnt menu items.

**Comment by dneukirchen** [comment]
@laoneo pls have a look at https://github.com/dneukirchen/tpl_tauristar/tree/jhtml/framework/libraries/cms/ui

**Comment by laoneo** [comment]
Cool idea, needs some refinements, but an approach I'm convinced would be the way to go. I would clean up the code and then open a PR with the content of the readme as pr text. We can discuss then further aspects directly on the code. Would make things easier.

**Comment by dneukirchen** [comment]
Thats the plan. I will create the PR next weekend (dont have time to finsish the code until then).

**Comment by mbabker** [comment]
There should not be an application configuration parameter, or template configuration parameter, defining the frontend framework in use.  This is a template level decision, not something that can or should be configured by the end user in the interface.  It's going to create more problems than it solves.  The end user should not be given the gun to shoot themselves in the foot by being able to tell Joomla that a Bootstrap 2 based template should load BS 3, 4, or any other framework.

**Comment by laoneo** [comment]
I agree here, but till we found a better place to let the template decide against which framework it is built to. We put it as hidden parameter https://github.com/Digital-Peak-Incubator/tpl_tauristar/blob/tpl_tauristar/templates/tauristar/templateDetails.xml#L67.

**Comment by mbabker** [comment]
Why does the template need to advertise its framework in any way?  Unless
you're intending for either core to ship a default set of layouts for a
wider range of CSS frameworks or introducing framework specific parameters
into the templating system as a whole, this is 100% unnecessary.

**Comment by laoneo** [comment]
The template decides against which fw it is built to. Where else you want to define it. Like that Joomla can introduce a new frontend framework, keeping the backwards compatibility.

**Comment by mbabker** [comment]
Why do you need to publicly advertise in any manner in the codebase what framework a template is built against?  This is what I'm having the hardest time understanding.

Core ships Bootstrap 2 right now and all of the media is loaded through the overridable loader methods in `JHtml`.  So if you need Bootstrap 3 JavaScript, you simply place the files at `templates/<template>/js/jui/bootstrap(.min).js` and you're all set.  There is no need to advertise the framework or introduce more deep rooted couplings into `JHtml`.

So your next issue is that the `JHtmlBootstrap` class is built against Bootstrap 2 but you need it to support Bootstrap 3.  `JHtml` supports method overrides, if you need it to do something different you simply register a different callable for the `bootstrap.*` keys to `JHtml`.  https://github.com/mbabker/bs3-demo is a very basic example of how to do this.  The plugin honestly needs some more efficient checks, but as a down and dirty demo, it just works.

So your next issue is you don't want to use Bootstrap at all but rather some other framework.  Tricker, but still possible to do.  Same concept with the override system, this time you're pointing the `JHtml` overrides to either a null method or a proxy to your framework system.  Of course, `JHtmlBootstrap` really shouldn't be getting called, so we need to ensure everything calling this class is in an overridable location.  A quick scan on the 3.7 branch shows that aside from the debug plugin, installer plugins, `JPagination`, `JToolbarButtonPopup`, and 3 form fields which aren't converted to using `JLayout` yet, all of the calls are either in layout files or `JHtml` helpers which are overridable.

**Comment by laoneo** [comment]
To have a painless upgrade to Joomla 4 with BS4, Joomla needs to support two frameworks at the same time. Otherwise every site which upgrades to Joomla 4 needs a new template. The extension developers need to maintain two versions of their extensions as it was the case with Joomla 1 and 1.5 and 2.5 and 3. This is not one click migration.

If every template developer with a BS3 based template in Joomla 3 needs to do what you propose, then it is just too complex. In praxis most templates nowadays have BS3 and it leads to many conflicts (try to open a bootstrap modal in your extension).

**Comment by mbabker** [comment]
Joomla core does NOT need to support multiple frameworks.  Hell, its core APIs should be so blissfully unaware of the framework in use there's no issue whatsoever for anyone to use whatever framework they want to use.

Fact is, every template developer HAS to do what I'm saying with the use of the override system to support BS3, or any other CSS framework.  In part because core has made a very bad decision with implementing tight internal coupling to a specific framework.  This is exactly WHY we have an override system!  The problem today is nobody does it then everyone bitches because the core system is so tightly coupled to Bootstrap 2; I just pointed out above that is NOT the case.

So in reality what you're saying is that Joomla going forward is going to be bloated as all hell with code and layouts to support Bootstrap 2, Bootstrap 3, Bootstrap 4 (which means shipping both a LESS and SASS compiler), Google's Material Design, Foundation, etc.

I do NOT see this project going forward in any way that's efficient.  Between the proposals from you and Cliff, I see Joomla becoming massively more complex for both end users AND developers (you're going to make me as an extension provider do a lot more UI related work to support your proposals, whereas today I support ONE thing only and that's whatever the default core template is and by following best practices I allow anyone using my extensions to make customizations to use whatever tools they want to use).  This is NOT the direction Joomla needs to go in.

**Comment by laoneo** [comment]
But what you are proposing is way more complex. When every template provider needs to override JHtmlBootstrap. To have a painless migration between two frameworks, you need to support two frameworks (not more) at the same time in Joomla.

I have to deal daily with this problem and Joomla doesn't try to solve it. Instead of repeating the same mistake with every major version. Or do you want to live with the fact that every bigger extensions is going to load it's namespaced version of frontend framework.

**Comment by mbabker** [comment]
The only way Joomla can solve the multiple framework issue is to either try to support every CSS framework in existence or build a framework agnostic application and implement a system that leaves ALL styling decisions at the templating level.  Someone is still going to get the shortend of the stick whether you try to support multiple frameworks or make core framework agnostic.

So yes, you do have to create a system where the providers override certain aspects of core.  This can be libraries that don't ship as part of the core package (think weblinks for a minute but as a library instead of a "normal" extension).  If a template depends on Bootstrap 2, install the library that provides that support; same for Bootstrap 3, or 4, or whatever else.  The whole point of a modular system like Joomla is to be able to do these types of things; not everything needs to ship as part of the core package but frankly nobody has the backbone to decouple all the optional stuff and work with this type of approach.

> Or do you want to live with the fact that every bigger extensions is going to load it's namespaced version of frontend framework.

To be frank, nothing this project is going to accomplish will cause the bigger extensions and template frameworks to change their workflows.  There is a reason Akeeba ships with its own rendering library (Strapper) and a third party rendering engine (Laravel's Blade).  There is a reason Gantry is as bloated as it is.  It's made worse by resources like the JED either publishing blog posts telling users how to break extensions or accepting extensions which have a sole purpose of breaking extensions (sorry to anyone who writes one of those extensions that lets you unload assets from JDocument, but your extensions are breaking other extensions in the process; the novelty of those extensions died during the 1.6-2.5 days when we made core stop arbitrarily loading MooTools).

**Comment by laoneo** [comment]
The library aspect is also on our radar. The only drawback there is that you have again duplicate code, which is something I try to avoid.

**Comment by mbabker** [comment]
You're going to have duplicated logic somewhere, its unavoidable.  Or you're going to have code that becomes super methods that tries to do a lot of stuff, which I'd say is much worse.

**Comment by laoneo** [comment]
Posted an RFC which addresses the duplicate code problem https://github.com/joomla/joomla-cms/pull/12016. Perhaps we will find a solution which works for all.
